panel_segmentation/lidar/pcd_data.py

```diff
@@ -0,0 +1,238 @@
+                # Create bbox
+                if self.polygon:
+                    bbox = shapely.contains_xy(
+                        projected_poly, x_coord, y_coord)
+                elif self.latitude and self.longitude:
+                    bbox = ((x_coord >= left) & (x_coord <= right) &
+                            (y_coord >= bottom) & (y_coord <= top))
+                # Filter out unnecessary classification and get only
+                # unclassified or building classification
+                # Classification info at:
+                # https://www.usgs.gov/ngp-standards-and-specifications/lidar-base-specification-tables
+                filter_class = (points.classification == 1) | (
+                    points.classification == 6)
+                filtered_points.append(points[bbox & filter_class].array)
+            if len(filtered_points) == 0:
+                print("Given latitude, longitude coordinates are not " +
+                      "within the LiDAR scan bounds.")
+            # Combine the points together and create a new LasData object
+            concatenated_points = np.concatenate(filtered_points)
+            final_point_record.array = concatenated_points
+            output_laz = laspy.LasData(header=f.header)
+            output_laz.points = final_point_record
+        return output_laz
+
+    def preprocessPcd(self, laz_data, nb_neighbors=10, std_ratio=0.5):
+        """
+        Preprocess pcd (point cloud data), which are 3D coordinate points
+        in the LiDAR data, and filters out noisy/outlier data.
+
+        Parameters:
+        -----------
+        laz_data: laspy.LasData
+            LasData LiDAR object that contains X,Y,Z coordinates.
+        nb_neighbors: int
+            Number of neighbors for outlier removal.
+            Defaulted to 10.
+        std_ratio : float
+            Standard deviation ratio threshold for outlier removal. The lower
+            the std_ratio, the more aggresive it is in remove outliers.
+            Defaulted to 0.5.
+
+        Returns:
+        --------
+        inlier_cloud: o3d.geometry.PointCloud or None
+            Preprocessed pcd with noise and outlier points removed.
+        """
+        # Ensure that the inputs are of the correct type
+        if not isinstance(laz_data, laspy.LasData):
+            raise TypeError(
+                "laz_data variable must be a laspy.LasData object.")
+        if not isinstance(nb_neighbors, int) or \
+                isinstance(nb_neighbors, bool):
+            raise TypeError("nb_neighbors variable must be of type int.")
+        if not isinstance(std_ratio, float):
+            raise TypeError("std_ratio variable must be of type float.")
+        # Make LasData object into PointCloud object
+        point_clouds = np.column_stack(
+            [laz_data["X"], laz_data["Y"], laz_data["Z"]])
+        pcd = o3d.geometry.PointCloud()
+        pcd.points = o3d.utility.Vector3dVector(point_clouds)
+        try:
```

Comment:
can we qualify this try-except, and turn it to an if-else, so failure is more explicit? ie only run if len(pcd.points) >0? This will explicitly handle the logic and if something else fails we'll know